route_following_plugin/src/route_following_plugin.cpp

```diff
@@ -568,15 +570,30 @@ void setManeuverLaneletIds(cav_msgs::Maneuver& mvr, lanelet::Id start_id, lanele
             ROS_DEBUG_STREAM("upcoming_lane_change_status_msg_map_.front().second.downtrack_until_lanechange: " <<static_cast<double>(upcoming_lane_change_status_msg_map_.front().second.downtrack_until_lanechange));
             upcoming_lane_change_status_pub_.publish(upcoming_lane_change_status_msg_map_.front().second); 
         }
-
-        auto current_lanelet = llts[0];
-        // if the current lanelet is not on the shortest path
-        if (shortest_path_set_.find(current_lanelet.id()) == shortest_path_set_.end())
-        {
-            returnToShortestPath(current_lanelet);
-   
+        
+        // Check if we need to return to route shortest path.
+        // Step 1. Check if another plugin aside from RFP has been in control
+        if (current_maneuver_plan_  != nullptr &&
+            GET_MANEUVER_PROPERTY(current_maneuver_plan_->maneuvers[0], parameters.planning_strategic_plugin) \
+            != planning_strategic_plugin_) {
+            // If another plugin may have brought us off shortest path, check our current lanelet
+            auto llts = wm_->getLaneletsFromPoint(current_loc, 10);                                          
+            // Remove any candidate lanelets not on the route
+            llts.erase(std::remove_if(llts.begin(), llts.end(),
+                [&](auto lanelet) -> bool { return !wm_->getRoute()->contains(lanelet); }));
+
+            // !!! ASSUMPTION !!!:
+            // Once non-route lanelets have been removed, it is assumed that our actual current lanelet is the only one that can remain.
+            // TODO: Verify that this assumption is true in all cases OR implement more robust logic to track current lanelet when there are overlaps
+            auto current_lanelet = llts[0];
```

Comment:
Adding a size check on this is good to avoid seg faults and could also provide some insight on if we ever have overlapping route lanelets while testing. At this point we are assuming it has size==1 right.

Reply:
Added size check and exception for size==0.